### DISCUSSION TITLE
what about lifetimes?

### DISCUSSION THREAD

**Discussion by voodookiidoo** [ratatui/ratatui#1487] [Ideas]
There are plenty of widgets and other stuff that requieres lifetime specifiying, like span, block , paragraph, text, etc. But all of it comes down to a single thing - Cow<'a, str>. I am new to rust, and i guess this was done for a higher performance. And even if so - maybe add it as a feature?
Just replace all Cow<str> with String and that's it!

**Comment by joshka** [comment]
The widgets were originally created as one shot commands, which were consumed, which meant having the option to pass in borrowed or owned strings was pretty useful. We've since made it possible to render widgets as references by implementing Widget on `&TheWidgetType`, which makes that less necessary. But there's a large amount of code that uses ratatui and this would probably the sort of change that would be fairly invasive / problematic to fix (just guessing here, I may be wrong).

The exact choice for types was made pre-fork days when this library was known as tui-rs, so if you're up for some code spelunking it's likely to be several years back.

An possible workaround to having many lifetimes in your own app code, is to make the lifetime of the widgets `'static`, which gives you essentially the same thing as String. You may find that in a bunch of places, you can omit the lifetime, or use an anonymous lifetime (`'_`) too. Give it a whirl with static, and see if this performs well enough for your purposes (check out cargo-flamegraph, and dhat for some ideas or a proper profiler).

I'm going to move this over to discussions as it's less of a problem than something that may spawn some ideas...
